src/librustc_trans/trans/common.rs

```diff
@@ -656,6 +657,14 @@ impl<'blk, 'tcx> BlockS<'blk, 'tcx> {
                                          self.fcx.param_substs,
                                          value)
     }
+
+    pub fn build(&self) -> Builder<'blk, 'tcx> {
+        // Create a fresh builder from the crate context.
+        let builder = self.ccx().builder();
+        // Set the builder's position to this block's end.
+        builder.position_at_end(self.llbb);
```

Comment:
> Each crate has only one builder, which BuilderRef points to.

You can create LLVM builders at will. The `BuilderRef` in `LocalCrateContext` is just a cached builder for the purposes of the old translator. No need to restrict ourselves only to the things available to the old translator.

> I'm worried the lack of repositioning could introduce bugs.

True, especially when there could be two different builders acting on the same basic block (i.e. bcx is passed to some `build::*` function), but the bugs are easily preventable by making `bcx` in `BlockBuilder` private and adding a method similar to this:

```rust
fn with_old_bcx<R, F: FnOnce(b: Block<...>) -> R> (&self, f: F) {
    f(self.bcx);
    self.builder.position_at_end(self.bcx.llbb);
}
```

> Also, keeping builder refs alongside blocks in the table would add at least several bytes of overhead for each block.

I don’t find that to be a problem. At any given point the translator is dealing with a small amount of blocks. I don’t think there’s any place in MIR trans where it would be necessary to have more than 2 or 3 `BlockBuilder`s alive at any given time. If you are worried about allocation overhead, dropped builders could be reused in some way as well.

> Otherwise, you'd need a different dance every time bcx is passed from MIR to common trans.

`bulder.with_old_bcx(|bcx| some_old_trans_function_wants_bcx(bcx, ...))` is much more trivial than sprinkling a `bcx.builder()` everywhere, where MIR trans wants to emit an instruction. Consider, that the old trans will be phased out in the future and we should adapt (once the old trans is gone) the common trans utilities to match the MIR trans model, and not the other way around.

All in all, while this PR indeed makes MIR’s use of LLVM builder more directly, in my opinion there’s much larger refactoring opportunities available. If you are not interested in exploring them, I’ll gladly r+ this if you change all `bcx.build()` to be `B(bcx)` (where `B` == `build::B`), because it does essentially the same thing as this method, but is shorter to type out.